docs: Announcing Rspack 1.0 (#7682)
web-infra-dev/rspack: website/docs/en/blog/announcing-1-0.mdx

---
date: 2024-06-29 16:00:00
---

# Announcing Rspack 1.0

> August 28, 2024

![](https://assets.rspack.dev/rspack/rspack-banner-v1-0.png)

---

> **Rspack 1.0 has been launched!**

Author: updated

> Rspack is a next-generation JavaScript bundler written in Rust, compatible with the webpack API and ecosystem, and is 10 times faster than webpack.

Author: I prefer the original 😄

> Rspack is a next-generation JavaScript bundler written in Rust, compatible with the webpack API and ecosystem, and is 10 times faster than webpack.

Reviewer: both works!

> Eighteen months ago, we open-sourced Rspack 0.1 and received substantial feedback and contributions from the community. During this time, 170 contributors have joined in the development of Rspack, submitting over 5000 pull requests and more than 2000 issues, which have helped Rspack release over 80 versions. And Rspack's weekly downloads on npm have exceeded 100,000 🎉.

Author: Worth another blog post 😄

![Rspack Stats](https://assets.rspack.dev/rspack/assets/rspack-v1-0-stats.png)

> Today Rspack has reached a new milestone - 1.0. This means that Rspack is production-ready, covers most of webpack's APIs and features, and is now prepared to support more users.

Author: Rspack has been battle-tested in community too, so I prefer the original

## New features

Since the release of 0.1, Rspack has introduced numerous important features and optimizations, including:

### Better performance

As a Rust-based bundler, performance has always been a core focus for Rspack. Since the release of Rspack 0.1, we have made numerous performance improvements, optimizing its performance for different scenarios, and supporting key features such as [lazy compilation](https://rspack.dev/config/experiments#experimentslazycompilation) to ensure better performance in large projects.

> Here is a comparison of build performance between Rspack 0.1 and Rspack 1.0 from the [benchmark](https://github.com/rspack-contrib/performance-compare). As shown, Rspack has significantly improved build performance while also adding many new features:

Reviewer: minor suggestion你们可以考虑下哈

![Rspack benchmark](https://assets.rspack.dev/rspack/assets/rspack-v1-0-benchmark.png)

> Note that the current architecture and implementation of Rspack still have significant room for optimization. After the 1.0 release, we plan to further improve the performance by several folds to better support large-scale application.

Reviewer 2: `several folds`  seems chinglish?

As the API support has gradually improved, Rspack has added support for more and more webpack plugins and loaders. Currently, Rspack is compatible with almost all loaders in the community. For the 50 most downloaded [webpack plugins](https://rspack.dev/guide/compatibility/plugin), more than 80% can be used in Rspack or have an alternative.

Building on this foundation, Rspack supports more libraries and frameworks, including React, Preact, Vue, Solid, Svelte, and NestJS. We would also like to thank the maintainers of many community plugins who have actively adapted their work for Rspack, such as [unplugin](https://github.com/unjs/unplugin) and [node-polyfill-webpack-plugin](https://www.npmjs.com/package/node-polyfill-webpack-plugin). Special thanks to [Alexander Akait](https://github.com/alexander-akait), one of the main maintainers of webpack, who helped us support many webpack loaders and plugins.

> We also hope to support and create more community plugins in the future, making the webpack and Rspack ecosystem even richer.

Author: Avoid repeating "Building on this foundation" as it has already been used in the previous paragraph

> We also hope to support and create more community plugins in the future, making the webpack and Rspack ecosystem even richer.

Reviewer: 好的
```suggestion
We also hope to support and create more community plugins to further enrich the webpack and Rspack ecosystem.
```

### Smaller bundle size

Rspack has consistently prioritized minimizing the bundle size of production builds. Since the release of 0.1, Rspack has gradually aligned its optimization capabilities with webpack, implementing features such as [split chunks](https://rspack.dev/plugins/webpack/split-chunks-plugin#splitchunksplugin), [tree shaking](https://rspack.dev/guide/optimization/tree-shaking), [scope hoisting](https://rspack.dev/config/optimization#optimizationconcatenatemodules) and [mangle exports](https://rspack.dev/config/optimization#optimizationmangleexports).

When a project migrates from webpack to Rspack, these features ensure that the bundle size remains the same as webpack while improving DX. In some scenarios, the output size of Rspack has even slightly outperformed webpack.

For example, in a real-world medium-sized web application, the bundle size of Rspack 1.0 was optimized from 6600KB to 5900KB compared to Rspack 0.1, which is equivalent to webpack. In the future, Rspack will continue to explore more advanced solutions to optimize bundle size.

### Support for Module Federation 2.0

> [Module Federation](https://module-federation.io/), as the most advanced micro-frontend architecture, has been widely adopted within the frontend ecosystem. The Rspack team has been working with the Module Federation team to develop Module Federation 2.0. The new version provides features such as dynamic TS type hints, Chrome devtools, runtime plugins, preloading. These features make Module Federation more suitable for use as a micro-frontend architecture in large-scale web applications.

Reviewer 2: `as the most advanced` this may cause some dispute? `as wildly used`?

Author: updated, removed `most advanced`

> [Module Federation](https://module-federation.io/) is a micro-frontend architectural pattern that is widely used in the ecosystem. The Rspack team has been working with the Module Federation team to develop Module Federation 2.0. The new version provides features such as dynamic TS type hints, Chrome devtools, runtime plugins, preloading. These features make Module Federation more suitable for use as a micro-frontend architecture in large-scale web applications.

Reviewer: The 会有点搞不清楚你在refer谁 (MF or Rspack)，This会很精确的说明你现在说的 `This new version` is Module Federation

Rspack also provides compatibility and support for Module Federation 1.0, making it easier for webpack projects to migrate.

### Stable API and new website

In 1.0, we have improved the stability of the configuration, JavaScript API, and plugin API. This ensures that higher-level tools and frameworks can more easily integrate with Rspack. We have also improved the guides and API documentation on the official website.

Rspack 1.0 also includes a brand new [homepage](https://rspack.dev/). Many thanks to designer Emily Jackson and team member Zack Jackson for their efforts in making this happen.

![Rspack Homepage](https://assets.rspack.dev/rspack/assets/rspack-v1-0-homepage.png)

## Why Rspack

Over the past two years, the community has seen the birth of several Rust-based bundlers, all of which have demonstrated remarkable performance. Rspack not only provides first-class performance. It also leads the community in terms of flexibility and compatibility.

The current goals of Rspack are:

- To help existing webpack projects progressively migrate to a high performance bundler, so that build performance is no longer a bottleneck for fast iterations.

> - Rspack is not just suitable for environments like Web and Node.js that we are familiar with; its goal is to cover all environments where JavaScript runs. This means that Rspack can easily support Deno, Electron, cross-platform applications, MiniApps, and any other JavaScript runtime.

Reviewer 2: ```suggestion
- Rspack is not just suitable for environments like Browser and Node.js that we are familiar with; its goal is to cover all environments where JavaScript runs. This means that Rspack can easily support Deno, Electron, cross-platform applications, MiniApps, and any other JavaScript runtime.
```

Reviewer 2: browser seems more accurate as environment?

Author: updated

> - Rspack is not just suitable for environments like browser and Node.js that we are familiar with; its goal is to cover all environments where JavaScript runs. This means that Rspack can easily support Deno, Electron, cross-platform applications, MiniApps, and any other JavaScript runtime.

Author: Most of them are already supported, we will add more doc in the future

> - Rspack is not just suitable for environments like browser and Node.js that we are familiar with; its goal is to cover all environments where JavaScript runs. This means that Rspack can easily support Deno, Electron, cross-platform applications, MiniApps, and any other JavaScript runtime.

Reviewer: ```suggestion
- Rspack is not just suitable for environments like browser and Node.js that we are familiar with; its goal is to cover all environments where JavaScript runs. This means that Rspack can easily support many JavaScript runtimes including Deno, Electron, cross-platform applications and MiniApps. We will add more documentation around them in the future.  
```

> - Rspack is not just suitable for environments like browser and Node.js that we are familiar with; its goal is to cover all environments where JavaScript runs. This means that Rspack can easily support Deno, Electron, cross-platform applications, MiniApps, and any other JavaScript runtime.

Author: Prefer the original

- We found that balancing "flexibility" and "out-of-the-box" in a single tool was a challenging task. Therefore, after open-sourcing Rspack, we developed a set of Rstack toolchains, including projects such as Rsbuild, Rspress, Rsdoctor, and Rslib, each targeting different use cases. For example, to reduce the complexity and high barriers to configuring Rspack, we provide Rsbuild for an out-of-the-box development experience.

### Rspack stack

![Rspack Stack](https://assets.rspack.dev/rspack/assets/rspack-v1-0-rstack.png)

Rstack is short for "Rspack Stack" and stands for the tech stack built around Rspack. It consists of the following tools:

- [Rspack](https://github.com/web-infra-dev/rspack): Focuses on implementing the high performance bundler, balancing performance and flexible configuration.

> - [Rsbuild](https://github.com/web-infra-dev/rsbuild): Focuses on building web applications, providing an out-of-the-box development experience.

Author: We will write another release note for Rsbuild 1.0 and provide detailed introduction.

> - [Rsbuild](https://github.com/web-infra-dev/rsbuild): Focuses on building web applications, providing an out-of-the-box development experience.

Reviewer: 好
```suggestion
- [Rsbuild](https://github.com/web-infra-dev/rsbuild): Focuses on building web applications, providing an out-of-the-box development experience. We will write another release note for Rsbuild 1.0 and provide more detailed introduction for it.
```

> - [Rsbuild](https://github.com/web-infra-dev/rsbuild): Focuses on building web applications, providing an out-of-the-box development experience.

Author: Prefer the original

- [Rslib](https://github.com/web-infra-dev/rslib): Focuses on building libraries, providing high quality ESM and CJS outputs.
- [Rspress](https://github.com/web-infra-dev/rspress) Focuses on generating static sites and supports MDX for building documentation sites and blogs.
- [Rsdoctor](https://github.com/web-infra-dev/rsdoctor) Focuses on build analysis, helping developers resolve build-related issues.

Together these tools make up the Rstack. We aim to provide a unified set of web development tools that deliver a top tier experience for both developers and users.

### Compatibility with webpack

Rspack 1.0 is designed to be compatible with webpack v5, which will help many projects using webpack to migrate smoothly to Rspack. However, this is a double-edged sword. To maintain compatibility with webpack, Rspack's design is limited, making it challenging to reach optimal solutions in terms of API, architecture, output format, and build performance. While maintaining compatibility with webpack, Rspack 1.0 also embraces modern web standards and aims for ultimate build performance.

> 

Author: Should be part of `Compatibility with webpack`

- For web standards, Rspack actively follows the evolution of modern web standards and the latest developments from TC39 and web standards. For example, Rspack already supports the use of Web Workers through [new Worker()](https://developer.mozilla.org/en-US/docs/Web/API/Worker/Worker), supports importing JSON modules through [Import Attributes](https://github.com/tc39/proposal-import-attributes), and supports importing CSS based on the [CSS Module Scripts](https://web.dev/articles/css-module-scripts) specification.

> - For performance, we have introduced many optimizations in 1.0. For example, if a JavaScript-side hook is not used, the Rust side will not invokecommunication with the JavaScript side. Also, Rspack performs lazy loading for many message objects. Even if the message object is large, if JavaScript only consumes a subset of its properties, Rspack will only transfer the consumed data, minimizing the communication overhead between Rust and JavaScript. And Rspack plans to provide even more lightweight hooks in the future to achieve more efficient communication between Rust and JavaScript.

Author: externally

In future major releases, Rspack will evolve based on the webpack API to better meet the needs of modern web development.

## Who's using Rspack

Since Rspack was open-sourced, many enterprises and developers have used Rspack in production. The weekly npm downloads of Rspack have also exceeded 100,000.

![Rspack downloads](https://assets.rspack.dev/rspack/assets/rspack-v1-0-downloads.png)

> Within ByteDance, Rspack's weekly downloads exceed 400,000, and over 1,000 web applications use Rspack, including TikTok, Douyin, Lark, Coze, and more. These projects have significantly improved build times and iteration efficiency by using Rspack. This has also helped us identify some early design issues with Rspack, prompting us to improve the architecture and strike a balance between migration cost, performance, and flexibility.

Author: Agree, moved to top

> We have also seen an increasing number of enterprise users starting to use Rspack, including Microsoft, Amazon, Alibaba, Intuit, Bit.dev, Discord, and others. We are excited that Rspack can help these enterprise users to achieve progressive migration, and we look forward to further cooperation and communication with more enterprises and developers in the future.

Author: done

![Who is using](https://assets.rspack.dev/rspack/assets/rspack-v1-0-who-is-using.png)

> ## How to use 1.0

Author: Title should be short

> ## How to use 1.0

Reviewer: 重要的事情还是要提一下吧不然他们upgrade然后break又要来找你lol

> ## How to use 1.0

Author: 下面已经介绍了如何迁移，并且不是所有项目都需要迁移，也会有很多新用户/新项目

> ## How to use 1.0

Reviewer: 好的

If you are using Rspack 0.7 or an earlier version, please note that 1.0 contains some breaking changes. We have prepared detailed documentation to help you upgrade. Please refer to: [Migration from Rspack 0.x](https://rspack.dev/guide/migration/rspack_0.x).

If you have never used Rspack before, please see [Quick Start](https://rspack.dev/guide/start/quick-start) to get started with Rspack. Also, feel free to give a star 🌟 to the [Rspack GitHub repository](https://github.com/web-infra-dev/rspack).

## What's next

Rspack 1.0 marks a new beginning. Following this release, the Rspack team will focus on the following goals:

- **Develop Rspack 1.x.** Rspack 1.x will iterate over 12 to 18 months, bringing more new features and improvements.

> - **Release Rsbuild 1.0.** It is based on Rspack 1.0 and supports [multi-environment builds](https://rsbuild.dev/guide/advanced/environments). Currently, Rsbuild has released version 1.0 RC, and the official release is expected in September.

Author: Release Candidate
https://en.wikipedia.org/wiki/Software_release_life_cycle

> - **Release Rsdoctor 1.0.** This release will improve support for Vue and provide report formats for CI/CD.

Author: Link added: https://github.com/web-infra-dev/rsdoctor/issues/408

- **Develop Rslib 0.x.** Rslib is a library building tool based on Rsbuild. See [Rslib repository](https://github.com/web-infra-dev/rslib) for more details.
- **Develop Rspress 2.0.** It will be based on React 19 and will improve some of the API designs. See [Rspress v2.0 planning](https://github.com/web-infra-dev/rspress/discussions/1105) for more details.

Here are some key features we plan to support in Rspack 1.x:

### Faster HMR

> Rspack can currently meet the performance requirements for most projects, but there is still significant room for performance optimization. During development, Rspack has already achieved nearly constant level incremental builds during the make phase. However, in the seal phase, some computations can still slow down as the project scales. Rspack will incrementally optimize the computations in the seal phase to keep the HMR time at a constant level.

Reviewer: small grammar fix

### Portable cache

> The evolution path of Rspack's caching capabilities follows a sequential implementation of memory cache, persistent cache, and portable cache. Currently, Rspack has implemented a memory cache that provides excellent HMR performance. The next step is to implement a persistent cache based on this foundation to address long cold startup times for large projects and to functionally align with webpack.

Reviewer: ```suggestion
The evolution path of Rspack's caching capabilities follows a sequential implementation of memory cache, persistent cache, and portable cache. Currently, Rspack has implemented a memory cache that provides excellent HMR performance. The next step is to implement a persistent cache based on this foundation to address long cold startup times for large projects,  and to not only building a solid, stable foundation by aligning with webpack's functionality, but also improve on top of this foundation.
```

Author: Prefer the original

> After that, we plan to continue implementing the portable cache. This means that Rspack's build cache will not only be persistent, but also portable across environments and machines. This will help teams make better use of the cache and lay the groundwork for distributed builds.

Reviewer: Is there any PR related to portable cache ? Didn't see it btw. 

Also using `the` means you are referring to a concept you have mentioned or at least people know it, but if no reference then we can make it bold (we are explaining it next line) ? 

```suggestion
After that, we plan to continue implementing **portable cache**. This means that Rspack's build cache will not only be persistent, but also portable across environments and machines. This will help teams make better use of the cache and lay the groundwork for distributed builds.
```

Author: Just planning, no PR yet

### TypeScript-based optimization

Currently, when Rspack processes TypeScript modules, it first converts them to JavaScript through a loader before further processing. This provides flexibility but also hinders further optimization of the build output. For example, developers need to use `enum` instead of `const enum`, but `enum` is difficult to optimize as a constant. In the future, we plan to treat TypeScript as a first-class citizen in Rspack, leveraging TypeScript's static information to provide more advanced compile-time optimization of the build output (such as [type-based property renaming](https://github.com/google/closure-compiler/wiki/Type-Based-Property-Renaming)).

### Stable Rust API

> Currently, upper-layer tools can use the JS API to integrate Rspack, which provides good extensibility. However, the communication overhead between Rust and JavaScript that limits the performance of Rspack. We also provide the SWC Wasm plugin to support extensions, but its performance is still slower than native languages.To provide upper-layer tools with more flexible integration options and better performance, we plan to expose Rspack's Rust API for integration.

Reviewer 2: ```suggestion
Currently, high level tools can use the JS API to integrate Rspack, which provides good extensibility. However, the communication overhead between Rust and JavaScript that limits the performance of Rspack. We also provide the SWC Wasm plugin to support extensions, but its performance is still slower than native languages.To provide upper-layer tools with more flexible integration options and better performance, we plan to expose Rspack's Rust API for integration.
```

Reviewer 2: upper-layer seems a direct translation from "上层"

> Currently, higher-level tools can use the JS API to integrate Rspack, which provides good extensibility. However, the communication overhead between Rust and JavaScript that limits the performance of Rspack. We also provide the SWC Wasm plugin to support extensions, but its performance is still slower than native languages.To provide higher-level tools with more flexible integration options and better performance, we plan to expose Rspack's Rust API for integration.

Reviewer: When you are saying higher-level tools, maybe name same examples like: 

```suggestion
Currently, higher-level tools (Rsbuild or build tools developed on top of Rspack) can use the JS API to integrate Rspack, which provides good extensibility. However, the communication overhead between Rust and JavaScript that limits the performance of Rspack. We also provide SWC Wasm plugin to support extensions, but its performance is still slower than native languages. To provide higher-level tools with more flexible integration options and better performance, we plan to expose Rspack's Rust API for integration.
```

Reviewer: Are you referring to SWC WASM plugin as in this one ? https://rspack.dev/plugins/webpack/enable-wasm-loading-plugin

or this one ? https://swc.rs/docs/usage/wasm

Author: The SWC one

Reviewer: ```suggestion
Currently, higher-level tools can use the JS API to integrate Rspack, which provides good extensibility. However, the communication overhead between Rust and JavaScript that limits the performance of Rspack. We also provide the [SWC Wasm plugin](https://swc.rs/docs/usage/wasm) to support extensions, but its performance is still slower than native languages.To provide higher-level tools with more flexible integration options and better performance, we plan to expose Rspack's Rust API for integration.
```

Author: done

### React Server Components support

At ByteDance, we have experimentally supported RSC (React Server Components) based on Rspack and validated it in a large web application. In the future, Rspack will provide first-class support for RSC, with more core features to make RSC easier to implement. For example, Rspack now supports the [layer](https://rspack.dev/config/experiments#experimentslayers) feature, which allows to build for multiple environments in a single run.

### Improved ESM output

> ESM is the standard for JavaScript modules. We are currently improving Rspack and webpack's support for ESM output and creating a library build tool based on Rspack called Rslib. This will allow developers to make better use of ESM's static analysis and tree-shaking when building npm packages.

Reviewer: ```suggestion
ESM is the standard for JavaScript modules. We are currently improving Rspack and webpack's support for ESM output and creating a library build tool based on Rspack called Rslib. This will allow developers to make better use of ESM's static analysis and tree-shaking when building npm packages. You can see an example on how Rslib is being used [in our Rslib examples here](https://github.com/web-infra-dev/rslib/blob/main/examples/react-component/rslib.config.ts).
```

Author: Rslib has been introduced above, so we won’t repeat it here.

- We also worked with the SWC team, contributing the Preact Refresh SWC plugin and fixing some transform and minify bugs in SWC. Thanks to [kdy](https://github.com/kdy1) for his review feedback.
- Rspack has embraced the [unplugin](https://github.com/unjs/unplugin) ecosystem and fully supports the unplugin API. Thanks to [sxzz](https://github.com/sxzz) for his review feedback and [antfu](https://github.com/antfu) for his remarkable creativity.

We are also excited to see Rspack being used or integrated into a wider ecosystem, including [Bazel](https://medium.com/@yanirmanor/why-moving-to-rspack-and-how-to-use-it-with-bazel-9f66139fe493), [Storybook](https://github.com/rspack-contrib/storybook-rsbuild), [Electron](https://github.com/noshower/electron-forge-plugin-rspack), and more.

Finally, we would like to thank all the developers who have contributed to the Rspack ecosystem ❤️:

![Rspack Contributors](https://assets.rspack.dev/rspack/assets/rspack-v1-0-contributors.png)

## FAQ

### What does the 1.0 release mean?

> The 1.0 release means that Rspack has implemented the core features of webpack and achieved API stability. Over the next 12 to 18 months, we will ensure the stability of the Rspack 1.x API so that developers can confidently build frameworks and tools on top of it. During the 1.x iteration, we may still find some inappropriate designs in Rspack. We will address these through progressive upgrades using [future flags](https://rspack.dev/config/experiments#experimentsrspackfuture).

Reviewer: ```suggestion
The 1.0 release means that Rspack has implemented the core features of webpack and achieved API stability. Over the next 12 to 18 months, we will ensure the stability of the Rspack 1.x API so that developers can confidently build frameworks and tools on top of it. During the 1.x iteration, we may still find some designs require polishing in Rspack. We will address these through progressive upgrades using [future flags](https://rspack.dev/config/experiments#experimentsrspackfuture).
```

Reviewer: 可以有自信， 中庸点，Things can be `polished`, `improved` 我们不用说自己不好

### Does Rspack follow semantic versioning?

Rspack follows semantic versioning (semver) and will not introduce breaking changes to the public API in minor or patch releases. Note that there are some exceptions:

> If your project has strict requirements for semantic versioning, you can pin Rspack to a minor version.

#### Experimental features

Rspack provides some experimental features that can be used via the [experiments](https://rspack.dev/config/experiments) config. In minor releases, Rspack may make changes to the APIs of these experimental features and provide detailed explanations of these changes in the release notes. So if you are using experimental features, please pay attention to the minor release notes.

#### SWC related features

> Rspack is built on SWC, which is currently in the pre-1.0 phase. To keep up with the fixes and improvements in SWC, we regularly update the SWC version. This may include some breaking changes in SWC or break some versions of the SWC Wasm plugins. In such cases, we will release a minor version of Rspack and add a note to the changelog.

Reviewer: ```suggestion
Rspack is built on SWC, which is currently in the pre-1.0 phase. To keep up with the fixes and improvements in SWC, we regularly update the SWC version. This may include some breaking changes in SWC or break some versions of the SWC Wasm plugins. In such cases, we will release a minor version of Rspack as soon as we can and add a note to the changelog.
```

Reviewer: 因为更新和迭代的速度其实很快，让读者减少担心

Author: 引入 breaking change 太频繁不是一个好的行为，实际迭代过程中我们也会控制 release minor 的频率
